src/lib.rs

```diff
@@ -0,0 +1,698 @@
+#[cfg(unix)]
+use unix::{memory_lock, memory_resize_and_lock};
+#[cfg(windows)]
+use windows::{memory_lock, memory_resize_and_lock, replace_set_size};
+use {
+    memtest::{MemtestError, MemtestKind, MemtestOutcome},
+    prelude::*,
+    rand::{seq::SliceRandom, thread_rng},
+    std::{
+        error::Error,
+        fmt,
+        time::{Duration, Instant},
+    },
+};
+
+mod memtest;
+mod prelude;
+
+#[derive(Debug)]
+pub struct Memtester {
+    test_kinds: Vec<MemtestKind>,
+    timeout: Duration,
+    mem_lock_mode: MemLockMode,
+    #[allow(dead_code)]
+    allow_working_set_resize: bool,
+    allow_multithread: bool,
+    allow_early_termination: bool,
+}
+
+// TODO: Replace MemtesterArgs with a Builder struct implementing fluent interface
+/// A set of arguments that define the behavior of Memtester
+#[derive(Debug)]
+pub struct MemtesterArgs {
+    /// How long should Memtester run the test suite before timing out
+    pub timeout: Duration,
+    /// Whether memory will be locked before testing and whether the requested memory size of
+    /// testing can be reduced to accomodate memory locking
+    /// If memory locking failed but is required, Memtester returns with error
+    pub mem_lock_mode: MemLockMode,
+    /// Whether the process working set can be resized to accomodate memory locking
+    /// This argument is only meaningful for Windows
+    pub allow_working_set_resize: bool,
+    /// Whether mulithreading is enabled
+    pub allow_multithread: bool,
+    /// Whether Memtester returns immediately if a test fails or continues until all tests are run
+    pub allow_early_termination: bool,
+}
+
+#[derive(Debug)]
+pub enum MemtesterError {
+    MemLockFailed(anyhow::Error),
+    Other(anyhow::Error),
+}
+
+#[derive(Debug)]
+pub struct MemtestReportList {
+    pub tested_mem_length: usize,
+    pub mlocked: bool,
+    pub reports: Vec<MemtestReport>,
+}
+
+#[derive(Debug)]
+pub struct MemtestReport {
+    pub test_kind: MemtestKind,
+    pub outcome: Result<MemtestOutcome, MemtestError>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum MemLockMode {
+    Resizable,
+    FixedSize,
+    Disabled,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct ParseMemLockModeError;
+
+/// The minimum memory length (in usize) for Memtester to run tests on
+/// On a 64-bit machine, this is the size of a page
+pub const MIN_MEMORY_LENGTH: usize = 512;
+
+#[derive(Debug)]
+struct MemLockGuard {
+    base_ptr: *const (),
```

Comment:
I think it makes the most sense to store this as `*mut usize`. It would eliminate a bunch of your casting below, and it's almost-always a bad idea to cast from a `*const` to a `*mut` (In this case, it's only okay because you had a `*mut` to begin with).